awx/main/scheduler/task_manager.py

```diff
@@ -255,72 +238,52 @@ def process_finished_workflow_jobs(self, workflow_jobs):
+                    name=task.name, pk=task.pk, timeout=task.timeout
+                )
+                logger.warning(timeout_message)
+                task.timed_out = True
+                task.status = 'failed'
+                task.send_approval_notification('timed_out')
+                task.websocket_emit_status(task.status)
+                task.job_explanation = timeout_message
+                task.save(update_fields=['status', 'job_explanation', 'timed_out'])
 
-        dependent_tasks = dependent_tasks or []
+    @timeit
+    def get_tasks(self):
+        return self.get_running_workflow_jobs()
 
-        task_actual = {
-            'type': get_type_for_model(type(task)),
-            'id': task.id,
-        }
-        dependencies = [{'type': get_type_for_model(type(t)), 'id': t.id} for t in dependent_tasks]
+    @timeit
+    def _schedule(self):
+        running_workflow_tasks = self.get_tasks()
+        if len(running_workflow_tasks) > 0:
+            self.process_finished_workflow_jobs(running_workflow_tasks)
 
-        task.status = 'waiting'
+            previously_running_workflow_tasks = running_workflow_tasks
+            running_workflow_tasks = []
+            for workflow_job in previously_running_workflow_tasks:
+                if workflow_job.status == 'running':
+                    running_workflow_tasks.append(workflow_job)
+                else:
+                    logger.debug('Removed %s from job spawning consideration.', workflow_job.log_format)
 
-        (start_status, opts) = task.pre_start()
-        if not start_status:
-            task.status = 'failed'
-            if task.job_explanation:
-                task.job_explanation += ' '
-            task.job_explanation += 'Task failed pre-start check.'
-            task.save()
-            # TODO: run error handler to fail sub-tasks and send notifications
-        else:
-            if type(task) is WorkflowJob:
-                task.status = 'running'
-                task.send_notification_templates('running')
-                logger.debug('Transitioning %s to running status.', task.log_format)
-                schedule_task_manager()
-            # at this point we already have control/execution nodes selected for the following cases
-            else:
-                task.instance_group = instance_group
-                execution_node_msg = f' and execution node {task.execution_node}' if task.execution_node else ''
-                logger.debug(
-                    f'Submitting job {task.log_format} controlled by {task.controller_node} to instance group {instance_group.name}{execution_node_msg}.'
-                )
-            with disable_activity_stream():
-                task.celery_task_id = str(uuid.uuid4())
-                task.save()
-                task.log_lifecycle("waiting")
+            self.spawn_workflow_graph_jobs(running_workflow_tasks)
```

Comment:
You know, @kdelee if you really want a hack with the `modified` time stamp that would help performance, I'd argue that you could update the `modified` time of a workflow job whenever a job inside of it finishes, and then this could only consider workflows with modified in the last 20 seconds for spawning or finishing. That would get rid of a lot of expensive reprocessing.

But I personally think the _most_ ideal solution would be for the job wrapup logic to trigger the workflow manager, _and pass in its workflow job id_ so you basically know you need to do something interesting, as opposed to running the graph logic on all workflows... every loop... which does nothing 99% of the time.

Reply:
@AlanCoding I had almost exactly the same thought last night! I think I know where we need to do this